### PR TITLE
First pass refactor of project explorer

### PR DESCRIPTION
**Pull request by SusanRatiLane**
This refactor removes some duplication and pushes project management closer to a model-view-controller separation of concerns.

There's a lot more that needs to be done here. Ideally, we wouldn't have so many objects being grabbed from anywhere via static methods that return their own instances. It's not totally clear to me that we need the event listeners at all -- it seems like, though I haven't proved it yet, that a redesign could cover all the listener functionality in the rpc callbacks.

There should be a role for an object ProjectManager as controller and ProjectList as view, but right now they overlap a lot. I think I removed the cases where they duplicated methods.

Any any rate, this is presented for heckling.

Addresses #1997 

Change-Id: I9e9a75a024b4ce9159164d587b1dcf27b8474a43

### CODE REVIEW

**Comment by SusanRatiLane** [comment]
> I haven't had a chance to review the code changes in depth, but I want to note two behavioral things that are different with this change compared with master (tested using code.appinventor nb181c):

I believe these issues are addressed.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java [hunk 1]
```diff
@@ -108,21 +105,23 @@ public void execute() {
...
-  private static class DeleteAction implements Command {
+  private class MoveToTrashAction implements Command {
     @Override
     public void execute() {
       Ode.getInstance().getEditorManager().saveDirtyEditors(new Command() {
         @Override
         public void execute() {
           List<Project> selectedProjects =
-              ProjectListBox.getProjectListBox().getProjectList().getSelectedProjects();
+                  ProjectListBox.getProjectListBox().getProjectList().getSelectedProjects();
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java]: Code style: 4 spaces only for continuation indent.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java [hunk 2]
```diff
@@ -131,60 +130,32 @@ public void execute() {
...
-    private boolean deleteConfirmation(List<Project> projects) {
-      String message;
-      GallerySettings gallerySettings = GalleryClient.getInstance().getGallerySettings();
-      if (projects.size() == 1) {
-        if (projects.get(0).isPublished()) {
-          message = MESSAGES.confirmDeleteSinglePublishedProjectWarning(projects.get(0).getProjectName());
-        } else {
-          message = MESSAGES.confirmMoveToTrashSingleProject(projects.get(0).getProjectName());
-        }
+  private boolean deleteConfirmation(List<Project> projects) {
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectToolbar.java]: Why was this moved up in scope? It is only ever called from within `MoveToTrashAction`.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java [hunk 3]
```diff
@@ -311,54 +319,62 @@ public void refreshTable(boolean needToSort) {
...
+            pw.publishedLabel.setText(PUBLISHED);
+          } else {
+            pw.publishedLabel.setText(NOT_PUBLISHED);
+          }
         }
       }
-
-      row++;
     }
+    table.resizeRows(row + 1);
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: OK, I'm not clear on what's going on under the hood here. 

The total size of the grid was set at line 314 of the original code. Then we loop through each project and populate the rows of the grid. We could do that because active projects and trashed projects were stored in separate but identical structures, so we knew the number of rows just from checking the size.

In the new code, I don't know at the start how large the grid will be because I am referencing the trash status of each project as I loop through. We still loop through the projects and pull the same information, resulting in a grid of the same size.

You're indicating that the setSize reuses existing rows, which wasn't obvious to me but makes sense. I can adjust this code to reuse existing grid rows if performance is a concern. Rather than setting the grid size before the loop, I'll have to check the size of the grid as I loop, add rows if I need them, and truncate anything unused at the end.

**Comment by SusanRatiLane** [comment]
@ewpatton I believe this is ready for another look.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/project/Project.java [hunk 4]
```diff
@@ -223,6 +225,73 @@ public void deleteNode(ProjectNode node) {
...
+                }
+              }
+            });
+  }
+
+  public void deleteFromTrash() {
+    Tracking.trackEvent(Tracking.PROJECT_EVENT,
+            Tracking.PROJECT_ACTION_DELETE_PROJECT_YA, getProjectName());
+    if (isPublished()) {
+      Ode.getInstance().getGalleryService().deleteApp(projectInfo.getGalleryId(),
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/project/Project.java]: This makes sense. (I flagged for another review, but I haven't addressed this yet.)

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java [hunk 5]
```diff
@@ -91,7 +89,7 @@ public ProjectList() {
     sortOrder = SortOrder.DESCENDING;
 
     // Initialize UI
-    table = new Grid(1, 5); // The table initially contains just the header row.
+    table = new Grid(3, 5); // The table initially contains just the header row.
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: Any reason as to why this was changed?

File: appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java [hunk 6]
```diff
@@ -186,10 +185,10 @@ public void updateMenuState(int numSelectedProjects, int numProjects) {
     boolean allowDelete = !isReadOnly && numSelectedProjects > 0;
     boolean allowExport = numSelectedProjects > 0;
     boolean allowExportAll = numProjects > 0;
+    fileDropDown.setItemEnabled(MESSAGES.trashProjectMenuItem(), allowDelete);
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java]: At some point I think we will want to have this menu item say "Delete from Trash" when in in the Trash view. Right now the only way to delete from the trash is to use the button as the menu item is completely neutered in this view.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java (hunk 5 above)
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: The Grid object only seems to have an insertRow(before) method, not an append method. insertRow requires an additional row at the end so that there is a valid row for rows to be inserted before.

The rows are then:
0: Header row
1: First row that will receive data when the grid is refreshed
2: Extra row so that I can insert rows before it

It's possible I'm somehow violating the intended paradigm of the Grid object because this seemed pretty awkward. If there's an append someplace I missed, please point me to it.

Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: Like I said, it works, but it just wasn't clear to me why it worked, especially given that the original code started with a single row.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java [hunk 7]
```diff
@@ -279,30 +277,40 @@ public void onClick(ClickEvent event) {
   // ProjectManagerEventListener interface that this is the
   // implementation of.
   public void refreshTable(boolean needToSort) {
+    if (Ode.getInstance().getCurrentView() == 8) {
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: I didn't tackle it in this refactor, but what's the value of getInstance() here? It seems like references between top-level objects are much more contorted than they need to be.

File: appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java (hunk 6 above)
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java]: That makes sense. I'll put in an issue for it.

File: appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java (hunk 5 above)
Comment (SusanRatiLane) [reply appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java]: refreshTable sets the Grid to a minimum of 3 rows, so it's not strictly necessary to do it in the initialization, but that's the smallest the Grid can be.

Probably worth mentioning this in a comment, since it's not intuitive.
